### PR TITLE
[Feature:Forum] Adding Upducks Feature

### PR DESCRIPTION
**Pull request by ArnavPawar**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant
* [ ] Screenshots are attached to Github PR if visual/UI changes were made

### What is the current behavior?
Added UI for upduck button that will allow users to like posts within a thread. 
This can be helpful for looking for the best post in a thread

### What is the new behavior?
image

### Other information?

You can test by going to different accounts and liking the button.

### CODE REVIEW

**Review by benjaminlee755 — CHANGES_REQUESTED** [review]
When I clicked into a discussion forum, I got a server error stating that the upducks_table doesn't exist.
image

**Review by DarthNyan — CHANGES_REQUESTED** [review]
Feature works for me. There is a change I'd like you to make about querying your data.

@benjaminlee755 did you run the up migration?

File: site/app/controllers/forum/ForumController.php [hunk 1]
```diff
@@ -1243,4 +1242,18 @@ public function showStats() {
         ksort($users);
         $this->core->getOutput()->renderOutput('forum\ForumThread', 'statPage', $users);
     }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/post/likes", methods={"POST"})
+     */
+    public function toggleLike() {
+        $output= [];
+        $output['type']=$this->core->getQueries()->toggleLikes($_POST['post_id'], $_POST['thread_id'], $_POST['current_user']);
```
Comment (cjreed121) [inline site/app/controllers/forum/ForumController.php:1251]: This will cause a server error if those keys don't exist. Could you make a small check before hand and return a json fail if they don't exist.

File: site/public/js/forum.js [hunk 2]
```diff
@@ -1251,6 +1251,68 @@ function modifyThreadList(currentThreadId, currentCategoriesId, course, loadFirs
...
+                // eslint-disable-next-line no-var
+                var json = JSON.parse(data);
+            }
+            catch (err) {
+                // eslint-disable-next-line no-undef
+                displayErrorMessage('Error parsing data. Please try again.');
+                return;
+            }
+            if (json['status'] === 'fail') {
+                // eslint-disable-next-line no-undef
```
Comment (cjreed121) [inline site/public/js/forum.js:1278]: You are ignoring quite a lot of eslint warnings. It would be better to fix these rather than ignore them.

**Review by cjreed121 — CHANGES_REQUESTED** [review]
Make sure you dump the modified database into the course SQL file. Also you have a lot of random unnecessary whitespace changes. Please revert those.

File: migration/migrator/migrations/course/20231113195556_create_upduck_table.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    :type database: migrator.db.Database
+    :param semester: Semester of the course being migrated
+    :type semester: str
+    :param course: Code of course being migrated
+    :type course: str
+    """
+
+    database.execute(
+    """
+        CREATE TABLE IF NOT EXISTS upducks_table (
```
Comment (williamjallen) [inline migration/migrator/migrations/course/20231113195556_create_upduck_table.py:21]: Typically we don't include "table" in our table names.  Perhaps `forum_upducks` would be more appropriate.

File: migration/migrator/migrations/course/20231113195556_create_upduck_table.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            thread_id integer NOT NULL,
+            FOREIGN KEY(post_id)
+                REFERENCES posts(id)
+                ON UPDATE CASCADE,
+            FOREIGN KEY(user_id)
+                REFERENCES users(user_id)
+                ON UPDATE CASCADE,
+            FOREIGN KEY(thread_id)
+                REFERENCES threads(id)
+                ON UPDATE CASCADE
```
Comment (williamjallen) [inline migration/migrator/migrations/course/20231113195556_create_upduck_table.py:33]: Do you also need `ON DELETE CASCADE` here?

**Review by williamjallen — CHANGES_REQUESTED** [review]
I'll have more requested changes for stylistic issues once this PR is closer to being finalized.

File: site/app/libraries/database/DatabaseQueries.php [hunk 5]
```diff
@@ -722,6 +722,50 @@ public function getCategoriesIdForThread($thread_id) {
...
+                $this->course_db->rollback();
+            }
+            return false;
+        }
+    }
+
+    public function getUpduckInfoForPosts(array $post_ids): array {
+        $resultDict = [];
+        foreach($post_ids as $post_id) {
+            $this->course_db->query('SELECT COUNT(*) AS cnt FROM forum_upducks WHERE post_id = ?', [$post_id]);
```
Comment (DarthNyan) [inline site/app/libraries/database/DatabaseQueries.php:752]: The idea behind moving the queries out of createPost is so that you won't run queries within a loop. Queries are slow and it's a lot faster to iterate over a large result from a query than to make a bunch of smaller queries.

File: site/app/libraries/database/DatabaseQueries.php [hunk 6]
```diff
@@ -722,6 +722,50 @@ public function getCategoriesIdForThread($thread_id) {
...
+            $resultDict[$post_id]= intval($this->course_db->row()['cnt']);
+        }
+        return $resultDict;
+    }
+
+    public function getUserLikesForPosts(array $post_ids, $current_user): array {
+        $result = [];    
+        foreach ($post_ids as $post){
+            $this->course_db->query('SELECT COUNT(*) AS cnt FROM forum_upducks WHERE post_id = ? AND user_id = ?', [$post, $current_user]);
```
Comment (DarthNyan) [inline site/app/libraries/database/DatabaseQueries.php:761]: Same here. Move this query outside of the loop, call createParameterList, and then do something like 
```suggestion
            $this->course_db->query('SELECT * FROM forum_upducks WHERE post_id IN ${parameters} AND user_id = ?', array_merge($post_ids, [$current_user]));
```
Then you can iterate over $this->course_db->rows() to format your output

**Review by DarthNyan — CHANGES_REQUESTED** [review]
Getting the query functions out of createPost is good, and this should help it merge with my PR. There's still some stuff I'd like you to do within the query functions, but this a good step.

I'm noticing when I refresh the page, I can keep clicking on posts I haven't upducked and the upduck number will keep going up and the button won't change to yellow until I refresh the page. Can you replicate and fix this?

**Review by VVVITH — APPROVED** [review]
Php lint test fixed, feature works as expected. Approve

File: site/app/views/forum/ForumThreadView.php [hunk 7]
```diff
@@ -474,6 +474,15 @@ public function generatePostList($currentThread, $posts, $unviewed_posts, $curre
         $merged_threads = $this->core->getQueries()->getMergedThreadIds($post_ids);
         $post_attachments = $this->core->getQueries()->getForumAttachments($post_ids);
 
+        $postIDs = [];
+        foreach ($posts as $post) {
+            array_push($postIDs, $post["id"]);
```
Comment (cjreed121) [inline site/app/views/forum/ForumThreadView.php:479]: Use different syntax

File: site/app/views/forum/ForumThreadView.php [hunk 8]
```diff
@@ -515,13 +524,18 @@ public function generatePostList($currentThread, $posts, $unviewed_posts, $curre
 
                         $post["author_user_group"] = $author_user_groups_map[$post["author_user_id"]];
 
+                        $boolLiked = true;
```
Comment (cjreed121) [inline site/app/views/forum/ForumThreadView.php:527]: unnecessary?

File: site/app/controllers/forum/ForumController.php [hunk 9]
```diff
@@ -673,7 +673,6 @@ public function alterPost($modify_type) {
             else {
                 $type = "post";
             }
-
```
Comment (cjreed121) [inline site/app/controllers/forum/ForumController.php:676]: unnecessary

File: site/public/js/forum.js [hunk 10]
```diff
@@ -1283,6 +1284,81 @@ function modifyThreadList(currentThreadId, currentCategoriesId, course, loadFirs
     });
 }
 
+function toggleLike(post_id, thread_id, current_user) {
+
+    // eslint-disable-next-line no-undef
```
Comment (cjreed121) [inline site/public/js/forum.js:1289]: Remove eslint ignores

**Review by cjreed121 — CHANGES_REQUESTED** [review]
Tested and works fine. Some code changes though.

File: site/app/views/forum/ForumThreadView.php (hunk 8 above)
Comment (ArnavPawar) [reply site/app/views/forum/ForumThreadView.php:527]: it needed a type for php liner so I did that to fix that error

File: site/public/js/forum.js (hunk 10 above)
Comment (ArnavPawar) [reply site/public/js/forum.js:1289]: I think I need it there. I checked other functions and every buildCourseUrl has this eslint ignore before they call the buildCourseUrl.

File: site/app/libraries/database/DatabaseQueries.php [hunk 11]
```diff
@@ -722,6 +722,76 @@ public function getCategoriesIdForThread($thread_id) {
...
+        }
+    }
+
+    /**
+     * @param int[] $post_ids
+     * @return int[]
+     */
+    public function getUpduckInfoForPosts(array $post_ids): array {
+        $placeholders = $this->createParameterList(count($post_ids));
+        $sql = "SELECT post_id, COUNT(*) AS cnt FROM forum_upducks WHERE post_id IN {$placeholders} GROUP BY post_id";
```
Comment (DarthNyan) [inline site/app/libraries/database/DatabaseQueries.php:755]: Queries look good and not iterated.

One thing I recently had to do with my queries is add a check for an empty post_ids array, and not executing the query if there are no post_ids, as trying to query an empty list will cause a server error (see #10055). Can you add similar checks here and to getUserLikesForPosts?

**Review by DarthNyan — DISMISSED** [review]
I haven't tested since I last reviewed, but I looked at the code and everything looks good except for one thing.

File: site/app/templates/forum/CreatePost.twig [hunk 12]
```diff
@@ -60,8 +60,12 @@
                     {% endif %}
                 </a>
             {% endif %}
-        {% endif %}
+        {% endif %} 
 
+        <button class="upduck-button text-decoration-none key_to_click" tabindex="0" onClick="toggleLike({{post.id}}, {{post.thread_id}}, '{{current_user}}')">
```
Comment (DarthNyan) [inline site/app/templates/forum/CreatePost.twig:65]: I'm not sure I like where the upduck is positioned as its between a whole bunch of other buttons and I think it would look better if it were off to one side. Maybe a discussion item.

**Review by DarthNyan — APPROVED** [review]
Looks good to me. Code and database queries look good and the feature works. I just have one thought about the positioning of the ducks.

File: site/app/libraries/database/DatabaseQueries.php [hunk 13]
```diff
@@ -722,6 +722,82 @@ public function getCategoriesIdForThread($thread_id) {
...
+            return "false";
+        }
+    }
+
+    /**
+     * @param int[] $post_ids
+     * @return int[]
+     */
+    public function getUpduckInfoForPosts(array $post_ids): array {
+        if (empty($post_ids)) {
```
Comment (DarthNyan) [inline site/app/libraries/database/DatabaseQueries.php:754]:
```suggestion
        if (count($post_ids) === 0) {
```
should fix the failing PHP lint test

File: site/app/libraries/database/DatabaseQueries.php [hunk 14]
```diff
@@ -722,6 +722,82 @@ public function getCategoriesIdForThread($thread_id) {
...
+        return $result;
+    }
+
+    /**
+     * @param int[] $post_ids
+     * @param string $current_user
+     * @return int[]
+     */
+    public function getUserLikesForPosts(array $post_ids, string $current_user): array {
+        if (empty($post_ids)) {
```
Comment (DarthNyan) [inline site/app/libraries/database/DatabaseQueries.php:778]:
```suggestion
        if (count($post_ids) === 0) {
```
Same here

**Review by arougey — APPROVED** [review]
Tested it and it works without issues. Potential improvement could be to have multiple upducks of different colors to count the number of likes by students and TAs/professors

**Review by cjreed121 — CHANGES_REQUESTED** [review]
I think it would also be beneficial for the front end to mark which way it is requesting to toggle. An example where this could be an issue if there was 2 pages of the same user, if they clicked to up-duck on each then they would have not upducked at all since they would cancel out. Adding a tiny bit of detail in the request can fix this.

File: migration/migrator/data/course_tables.sql [hunk 15]
```diff
@@ -887,6 +887,17 @@ CREATE TABLE public.forum_posts_history (
...
 
 
+--
+-- Name: forum_upducks; Type: TABLE; Schema: public; Owner: -
+--
+
+CREATE TABLE public.forum_upducks (
+    post_id integer NOT NULL,
+    user_id character varying(255) NOT NULL,
+    thread_id integer NOT NULL
```
Comment (cjreed121) [inline migration/migrator/data/course_tables.sql:897]: Make sure to update this file with the changed migration

File: site/app/libraries/database/DatabaseQueries.php [hunk 16]
```diff
@@ -722,6 +722,82 @@ public function getCategoriesIdForThread($thread_id) {
...
+
+            if (isset($this->course_db->rows()[0])) {
+                // If a row with the same values exists, delete it
+                $this->course_db->query("DELETE FROM forum_upducks WHERE post_id = ? AND user_id = ? AND thread_id = ?", [$post_id, $current_user, $thread_id]);
+                $result = 'unlike';
+            }
+            else {
+                // If no row with the same values exists, insert the new row
+                $this->course_db->query("INSERT INTO forum_upducks (post_id, user_id, thread_id) VALUES (?, ?, ?)", [$post_id, $current_user, $thread_id]);
+                $result = 'like';
```
Comment (cjreed121) [inline site/app/libraries/database/DatabaseQueries.php:737]: Remove thread_id from the queries since it won't be in the database

**Review by cjreed121 — CHANGES_REQUESTED** [review]

**Comment by JoJo-ESC** [comment]
After installing the branch I can't seem to actually find the up ducks button every time I try the up duck doesn't show up is there something I have to enable in settings?

**Review by RayReyRam — COMMENTED** [review]
Testing on my machine works fine, the feature doesn't have any apparent bugs, though I find up ducking slow, when clicked on it feels there is a pause and when spammed on, its function is delayed. I also think its a good idea to have some sort of animation or apparent change when you are hovering over the duck.

**Review by JoJo-ESC — CHANGES_REQUESTED** [review]
Functionality is there and is close to being finished. Cant remember if we talked about the duck changing when an instructor likes something but that should be implemented. Additionally, as previous comments stated, the time it takes to actually like the duck is a bit too lengthy and should be shortened if possible

**Comment by ArnavPawar** [comment]
> Functionality is there and is close to being finished. Cant remember if we talked about the duck changing when an instructor likes something but that should be implemented. Additionally, as previous comments stated, the time it takes to actually like the duck is a bit too lengthy and should be shortened if possible

I cant seem to get a delay when up ducking a post can you give more info on what you did to get the delay?

**Comment by RayReyRam** [comment]
> > Functionality is there and is close to being finished. Cant remember if we talked about the duck changing when an instructor likes something but that should be implemented. Additionally, as previous comments stated, the time it takes to actually like the duck is a bit too lengthy and should be shortened if possible
> 
> I cant seem to get a delay when up ducking a post can you give more info on what you did to get the delay?

I've attached a video below, The delay I was referring to was when the duck is clicked multiple times, and then stop clicking there is a delay in catching up. When looking at other platforms like YouTube spamming the like button this doesnt happen.

https://github.com/Submitty/Submitty/assets/148298134/8e120671-13a7-441c-9d90-150b865cde00

**Review by JoJo-ESC — APPROVED** [review]
After talking to Arnav in meeting he seemed to have fix any delay on actually liking the ducks itself. after a cursory code glance the code looks good.

**Review by williamlin6803 — APPROVED** [review]
Tested it and it works without issues. I tried my best to make sense of the code and it looks correct

**Review by RayReyRam — APPROVED** [review]
Talked in class, issue where spamming the duck causing latency, as Chris says this is an issue with the virtual machine itself and not an actual problem with the code. The program looks good runs well and I cant seem to break it. I approve.

File: site/app/views/forum/ForumThreadView.php (hunk 8 above)
Comment (cjreed121) [reply site/app/views/forum/ForumThreadView.php:523]: What was the linting error?

File: site/app/libraries/database/DatabaseQueries.php [hunk 17]
```diff
@@ -722,6 +722,89 @@ public function getCategoriesIdForThread($thread_id) {
         return $categories_list;
     }
 
+    public function toggleLikes(int $post_id, string $current_user, bool $action): string {
```
Comment (cjreed121) [inline site/app/libraries/database/DatabaseQueries.php:725]: Add a comment saying what $action means (i.e. what true/false represents). It is not clear when looking at just this code. Or maybe rename $action to a more descriptive name. The code below is a hard to follow which cases handle what.

File: site/public/js/forum.js [hunk 18]
```diff
@@ -1284,6 +1285,81 @@ function modifyThreadList(currentThreadId, currentCategoriesId, course, loadFirs
...
+                displayErrorMessage('Error parsing data. Please try again.');
+                return;
+            }
+            if (json['status'] === 'fail') {
+                // eslint-disable-next-line no-undef
+                displayErrorMessage(json['message']);
+                return;
+            }
+            json=json['data'];
+            //window.alert(json.type);
```
Comment (cjreed121) [inline site/public/js/forum.js:1318]: Remove this

**Review by cjreed121 — CHANGES_REQUESTED** [review]
Code looks good for most part. Just a couple things to clean up so functionality is clear to future developers.

File: site/public/js/forum.js (hunk 18 above)
Comment (ArnavPawar) [reply site/public/js/forum.js:1318]: what do you mean this?

Comment (cjreed121) [reply site/public/js/forum.js:1318]: The comment

File: site/app/views/forum/ForumThreadView.php (hunk 8 above)
Comment (ArnavPawar) [reply site/app/views/forum/ForumThreadView.php:523]: I think it was something about not being initialized or not being declared anywhere.

Comment (cjreed121) [reply site/app/views/forum/ForumThreadView.php:523]: Can you remove it and generate that linting error? We definitely should not need that line.

**Review by cjreed121 — APPROVED** [review]
